Explain the theoretical Simple Linear Regression model in your own words by describing its components (of predictor and outcome variables, slope and intercept coefficients, and an error term) and how they combine to form a sample from normal distribution; then, (working with a ChatBot if needed) create python code explicitly demonstrating your explanation using numpy and scipy.stats

Simple Linear Regression is a statistical model that describes a linear relationship between a dependent variable (often called the outcome or response variable) and one independent variable (also known as the predictor or explanatory variable). The model has three main components:

Predictor Variable (X): This is the variable that is manipulated or measured to predict or explain changes in the outcome variable.

Outcome Variable (Y): This is the variable that is being predicted or explained based on the predictor variable.

Slope Coefficient (β₁): This represents the change in the expected value of the outcome variable for a one-unit change in the predictor variable, holding all other factors constant.

Intercept Coefficient (β₀): This is the expected value of the outcome variable when the predictor variable is equal to zero.

Error Term (ε): This represents the variability or randomness in the outcome variable that cannot be explained by the predictor variable. It captures other factors that influence the outcome variable.

The equation for a simple linear regression model is:

Y
i
=
β
0
+
β
1
X
i
+
ϵ
i



Where:

Y
i is the outcome for the
i-th observation,

β
0 is the intercept,

β
1
​
  is the slope,

X i
​ is the value of the predictor variable for the
i-th observation,

ϵ i
​
  is the error term for the
i-th observation, which is assumed to be normally distributed with mean 0 and constant variance.
𝜖
 is normally distributed,
𝜖
∼
𝑁
(
0
,
𝜎
^2
)

In [ ]:
import numpy as np
import scipy.stats as stats
import plotly.graph_objects as go

# Parameters
n = 100         # Number of samples
beta0 = 2       # Intercept
beta1 = 3       # Slope
sigma = 1.5     # Standard deviation of the errors

# Fixed predictor values
x = np.linspace(0, 10, n)

# Error terms sampled from a normal distribution
errors = np.random.normal(0, sigma, n)

# Outcome variable y
y = beta0 + beta1 * x + errors

# Plotting the results
fig = go.Figure()
fig.add_trace(go.Scatter(x=x, y=y, mode='markers', name='Data'))
fig.add_trace(go.Scatter(x=x, y=beta0 + beta1 * x, mode='lines', name='Regression Line'))
fig.update_layout(title='Simple Linear Regression Simulation',
                  xaxis_title='Predictor Variable (x)',
                  yaxis_title='Outcome Variable (y)',
                  showlegend=True)
fig.show()


2. Use a dataset simulated from your theoretical Simple Linear Regression model to demonstrate how to create and visualize a fitted Simple Linear Regression model using pandas and import statsmodels.formula.api as smf

In [ ]:
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import plotly.express as px

# Set the random seed for reproducibility
np.random.seed(0)

# Generate random data for the predictor variable X
X = np.random.uniform(-1, 1, 100)

# Define the slope and intercept coefficients
beta_1 = 2  # Slope
beta_0 = 3  # Intercept

# Generate the true outcome variable Y without error
Y_true = beta_0 + beta_1 * X

# Generate the error term ε, assumed to be normally distributed with mean 0 and standard deviation 0.5
epsilon = np.random.normal(0, 0.5, 100)

# Generate the observed outcome variable Y with error
Y_obs = Y_true + epsilon

# Combine X and Y into a pandas DataFrame
df = pd.DataFrame({'x': X, 'y': Y_obs})

# Specify and fit the OLS regression model
model_data_specification = smf.ols("y~x", data=df)
fitted_model = model_data_specification.fit()

# Print the model summary
print(fitted_model.summary())

# Add a column for the legend
df['Data'] = 'Data'

fig = px.scatter(df, x='x',  y='y', color='Data',
                 trendline='ols', title='y vs. x')

# This is essentially what above `trendline='ols'` does
fig.add_scatter(x=df['x'], y=fitted_model.fittedvalues,
                line=dict(color='blue'), name="trendline='ols'")

fig.show()

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.841
Model:                            OLS   Adj. R-squared:                  0.839
Method:                 Least Squares   F-statistic:                     517.0
Date:                Thu, 31 Oct 2024   Prob (F-statistic):           7.18e-41
Time:                        14:50:13   Log-Likelihood:                -72.200
No. Observations:                 100   AIC:                             148.4
Df Residuals:                      98   BIC:                             153.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.0953      0.051     61.245      0.0

3. Add the line from Question 1 on the figure of Question 2 and explain the difference between the nature of the two lines in your own words

The red dashed line (theoretical model) represents the underlying relationship between the predictor variable X and the outcome variable Y without any random noise. It is generated using the true slope (beta_1) and intercept (beta_0) coefficients. The true line is a straight line that passes through all the points in the simulated data without any error. The blue line (fitted model) represents the line estimated from the sample data. It is generated using the estimated slope and intercept coefficients obtained from fitting the OLS regression model to the observed data. The fitted regression line may not pass through all the data points exactly because it accounts for the random error term in the data.The differences between these lines arise due to random sampling variation in the data, which influences the slope and intercept estimated by the regression.

Explanation of the Difference Between the Two Lines:

**True Line:** This line represents the underlying relationship between the predictor variable \( X \) and the outcome variable \( Y \) without any random noise. It is generated using the true slope (\( \beta_1 \)) and intercept (\( \beta_0 \)) coefficients. The true line is a straight line that passes through all points in the simulated data without any error.

**Fitted Regression Line:** This line is the best linear approximation of the observed data points. It is generated using the estimated slope and intercept coefficients obtained from fitting the OLS regression model to the observed data. The fitted regression line may not pass through all data points exactly because it accounts for the random error term in the data.

The true line is a theoretical construct that represents the actual relationship between the variables, while the fitted regression line is an empirical estimate based on the observed data. The difference between these two lines illustrates the impact of random noise on the observed data and how the regression model attempts to find the best fit despite this noise.

In [ ]:
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import plotly.express as px

# Set the random seed for reproducibility
np.random.seed(0)

# Generate random data for the predictor variable X
X = np.random.uniform(-1, 1, 100)

# Define the slope and intercept coefficients
beta_1 = 2  # Slope
beta_0 = 3  # Intercept

# Generate the true outcome variable Y without error
Y_true = beta_0 + beta_1 * X

# Generate the error term ε, assumed to be normally distributed with mean 0 and standard deviation 0.5
epsilon = np.random.normal(0, 0.5, 100)

# Generate the observed outcome variable Y with error
Y_obs = Y_true + epsilon

# Combine X and Y into a pandas DataFrame
df = pd.DataFrame({'x': X, 'y': Y_obs})

# Specify and fit the OLS regression model
model_data_specification = smf.ols("y~x", data=df)
fitted_model = model_data_specification.fit()

# Print the model summary
print(fitted_model.summary())

# Add a column for the legend
df['Data'] = 'Data'

# Create a scatter plot with a regression trendline
fig = px.scatter(df, x='x', y='y', color='Data', trendline='ols', title='y vs. x')

# Add the regression line manually for clarity
fig.add_scatter(x=df['x'], y=fitted_model.fittedvalues,
                line=dict(color='blue'), name="OLS Trendline")

# Add the true line without error
fig.add_scatter(x=df['x'], y=Y_true,
                line=dict(color='red', dash='dash'), name="True Line")

# Show the figure
fig.show()  # Use this for GitHub and MarkUs submissions

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.841
Model:                            OLS   Adj. R-squared:                  0.839
Method:                 Least Squares   F-statistic:                     517.0
Date:                Thu, 31 Oct 2024   Prob (F-statistic):           7.18e-41
Time:                        14:51:06   Log-Likelihood:                -72.200
No. Observations:                 100   AIC:                             148.4
Df Residuals:                      98   BIC:                             153.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.0953      0.051     61.245      0.0

4. Explain how fitted_model.fittedvalues are derived on the basis of fitted_model.summary().tables[1] (or more specifically fitted_model.params or fitted_model.params.values)

The `fitted_model.fittedvalues` attribute in a linear regression model represents the predicted values of the dependent variable (outcome variable) based on the linear equation derived from the fitted model. These predicted values are calculated using the estimated coefficients (slope and intercept) obtained from the data. Let's break down how these fitted values are derived from the model parameters.

### Understanding `fitted_model.params` and `fitted_model.params.values`

When you fit a linear regression model using `statsmodels`, the estimated parameters (coefficients) are stored in `fitted_model.params`. For a simple linear regression model, this will include:

- The intercept term (often labeled as `Intercept` in the output).
- The slope coefficient for the predictor variable (in this case, `x`).

`fitted_model.params.values` provides the numerical values of these parameters in an array, excluding the intercept. So, for a simple linear regression model, `fitted_model.params.values` will contain only the slope coefficient.

### Derivation of `fitted_model.fittedvalues`

The `fitted_model.fittedvalues` are calculated by applying the linear equation of the fitted model to each set of predictor values in the dataset. The general form of the linear equation for a simple linear regression model is:

\[ \hat{Y} = \beta_0 + \beta_1 X \]

Where:
- \( \hat{Y} \) are the predicted values of the outcome variable.
- \( \beta_0 \) is the estimated intercept.
- \( \beta_1 \) is the estimated slope coefficient.
- \( X \) is the value of the predictor variable.

`fitted_model.fittedvalues` is essentially a vector of \( \hat{Y} \) values calculated for each observation in the dataset using the above equation and the estimated parameters from `fitted_model.params`.

### Example Illustration

Let's say from `fitted_model.params`, you get:
- Intercept (\( \beta_0 \)) = 3
- Slope for `x` (\( \beta_1 \)) = 2

And your dataset `df` contains the following values for `x`:
- 0.5, 1.2, -0.3, etc.

The predicted values (\( \hat{Y} \)) for these `x` values would be calculated as follows:
- For \( x = 0.5 \): \( \hat{Y} = 3 + 2 \times 0.5 = 4 \)
- For \( x = 1.2 \): \( \hat{Y} = 3 + 2 \times 1.2 = 5.4 \)
- For \( x = -0.3 \): \( \hat{Y} = 3 + 2 \times (-0.3) = 2.4 \)

These calculated \( \hat{Y} \) values are what `fitted_model.fittedvalues` contains.

### Contrast with Theoretical Model

In contrast to the theoretical (true) Simple Linear Regression model, which uses the true parameters (\( \beta_0 \) and \( \beta_1 \)) to predict \( Y \), the fitted model uses estimates of these parameters derived from the data. The fitted model's predictions (\( \hat{Y} \)) are therefore an approximation of the true relationship, accounting for the variability and noise present in the data.

This process of deriving `fitted_model.fittedvalues` from `fitted_model.params` or `fitted_model.params.values` is central to understanding how linear regression models make predictions and is a key aspect of statistical modeling and inference.


5. Explain concisely in your own words what line is chosen for the fitted model based on observed data using the "ordinary least squares" method (as is done by trendline='ols' and smf.ols(...).fit()) and why it requires "squares"

The "ordinary least squares" (OLS) method used for fitting a model line to observed data works by minimizing the sum of the squares of the residuals. Here's a clear breakdown of the concept:

1. **Definition of Residuals**: Residuals are the differences between the observed values and the values predicted by the model. In your provided Python script, these are visually represented by red dashed lines between the actual data points (observed values) and the blue line (predicted values or fitted line).

2. **Purpose of Squaring Residuals**: Squaring each residual does two main things:
   - **Eliminates Negative Values**: Residuals can be negative if the predicted value is greater than the observed value. Squaring each residual ensures that all values are positive, making it possible to sum them.
   - **Emphasizes Larger Errors**: Squaring gives more weight to larger residuals. This is because the square of a larger number is much greater than the square of a smaller number. Therefore, OLS particularly penalizes larger discrepancies between the predicted and actual values, which helps in fitting a line that is as close as possible to most of the data points.

3. **Minimization Objective**: OLS seeks to find the parameters (in simple linear regression, these are the slope and intercept) that minimize the total sum of these squared residuals. This minimization process results in the best-fit line that has the least overall deviation from the actual data points when these deviations are considered collectively and squared.

4. **Visualization and Interpretation**:
   - In your script, the actual data points are plotted along with the OLS trendline.
   - The residuals (differences between observed and fitted values) are highlighted with vertical red dashed lines, directly showing how each data point contributes to the total sum of squares that OLS aims to minimize.
   - The trendline ('ols') shown in blue represents the best linear approximation of the data according to the OLS criteria of minimizing the sum of squared residuals.

5. **Why "Squares" in OLS**: The squaring of residuals in OLS is not only mathematically convenient (it simplifies the computation and derivatives used in optimization), but it also has practical implications in terms of focusing on reducing larger errors more than smaller ones, which typically results in a more accurate and robust model fit to the data.

By focusing on minimizing these squared residuals, OLS effectively adjusts the slope and intercept of the line to find the optimal linear relationship that best describes the observed data, ensuring that the model is as accurate as possible given the linear constraints. This method of fitting is particularly useful because it provides a clear and quantifiable objective for optimization and is computationally straightforward to implement.

6. Confirm that the following explain what the two np.corrcoef... expressions capture, why the final expression can be interpreted as "the proportion of variation in (outcome) y explained by the model (fitted_model.fittedvalues)", and therefore why fitted_model.rsquared can be interpreted as a measure of the accuracy of the model

1.fitted_model.rsquared

2.np.corrcoef(y,x)[0,1]**2

3.np.corrcoef(y,fitted_model.fittedvalues)[0,1]**2

4.1-((y-fitted_model.fittedvalues)**2).sum()/((y-y.mean())**2).sum()


Let's break down each of these expressions and understand what they capture in the context of a Simple Linear Regression model.

### 1. `fitted_model.rsquared`

`fitted_model.rsquared` is a statistical measure called R-squared (or the coefficient of determination). It represents the proportion of the variance in the dependent variable (outcome variable `y`) that is predictable from the independent variable(s) (in this case, `x`). R-squared is calculated as:

\[ R^2 = 1 - \frac{\text{SS}_{\text{res}}}{\text{SS}_{\text{tot}}} \]

Where:
- \(\text{SS}_{\text{res}}\) is the sum of squares of residuals, which is the sum of squared differences between the observed values (`y`) and the predicted values (`fitted_model.fittedvalues`).
- \(\text{SS}_{\text{tot}}\) is the total sum of squares, which is the sum of squared differences between the observed values (`y`) and the mean of `y`.

R-squared ranges from 0 to 1, with higher values indicating a better fit of the model to the data.

### 2. `np.corrcoef(y, x)[0,1]**2`

This expression calculates the square of the correlation coefficient between `y` and `x`. The correlation coefficient measures the strength and direction of the linear relationship between two variables. Squaring this value gives the coefficient of determination, which is the proportion of the variance in `y` that can be explained by `x` without any model fitting, just based on their linear relationship.

### 3. `np.corrcoef(y, fitted_model.fittedvalues)[0,1]**2`

This expression calculates the square of the correlation coefficient between `y` and the predicted values (`fitted_model.fittedvalues`). This is another way to estimate the proportion of the variance in `y` that is explained by the model. A high value close to 1 indicates that the predicted values are highly correlated with the actual values, suggesting a good fit.

### 4. `1-((y-fitted_model.fittedvalues)**2).sum()/((y-y.mean())**2).sum()`

This expression is another way to calculate R-squared. It is derived directly from the definition of R-squared:

\[ R^2 = 1 - \frac{\sum (y - \hat{y})^2}{\sum (y - \bar{y})^2} \]

Where:
- \(\sum (y - \hat{y})^2\) is the sum of squared residuals.
- \(\sum (y - \bar{y})^2\) is the total sum of squares.

This expression calculates the proportion of the variance in `y` that is explained by the model. It does so by comparing the variability of `y` around its mean to the variability of `y` around the predicted values.

### Interpretation

All four expressions capture the same underlying concept: the proportion of variation in the outcome variable `y` that is explained by the model. They all provide measures of how well the model fits the data, with higher values indicating a better fit. `fitted_model.rsquared` is a direct measure of this from the fitted model, while the other expressions provide alternative ways to estimate the same concept using the data and the predictions.

Thus, `fitted_model.rsquared` can indeed be interpreted as a measure of the accuracy of the model, as it quantifies how much of the variability in the outcome variable is accounted for by the model. This makes R-squared a useful metric for assessing the goodness of fit of a regression model.


7.Indicate a couple of the assumptions of the Simple Linear Regression model specification do not seem compatible with the example data below

In [ ]:
import pandas as pd
from scipy import stats
import plotly.express as px
from plotly.subplots import make_subplots

# This data shows the relationship between the amount of fertilizer used and crop yield
data = {'Amount of Fertilizer (kg) (x)': [1, 1.2, 1.4, 1.6, 1.8, 2, 2.2, 2.4, 2.6,
                                          2.8, 3, 3.2, 3.4, 3.6, 3.8, 4, 4.2, 4.4,
                                          4.6, 4.8, 5, 5.2, 5.4, 5.6, 5.8, 6, 6.2,
                                          6.4, 6.6, 6.8, 7, 7.2, 7.4, 7.6, 7.8, 8,
                                          8.2, 8.4, 8.6, 8.8,9, 9.2, 9.4, 9.6],
        'Crop Yield (tons) (y)': [18.7, 16.9, 16.1, 13.4, 48.4, 51.9, 31.8, 51.3,
                                  63.9, 50.6, 58.7, 82.4, 66.7, 81.2, 96.5, 112.2,
                                  132.5, 119.8, 127.7, 136.3, 148.5, 169.4, 177.9,
                                  186.7, 198.1, 215.7, 230.7, 250.4, 258. , 267.8,
                                  320.4, 302. , 307.2, 331.5, 375.3, 403.4, 393.5,
                                  434.9, 431.9, 451.1, 491.2, 546.8, 546.4, 558.9]}
df = pd.DataFrame(data)
fig1 = px.scatter(df, x='Amount of Fertilizer (kg) (x)', y='Crop Yield (tons) (y)',
                  trendline='ols', title='Crop Yield vs. Amount of Fertilizer')

# Perform linear regression using scipy.stats
slope, intercept, r_value, p_value, std_err = \
    stats.linregress(df['Amount of Fertilizer (kg) (x)'], df['Crop Yield (tons) (y)'])
# Predict the values and calculate residuals
y_hat = intercept + slope * df['Amount of Fertilizer (kg) (x)']
residuals = df['Crop Yield (tons) (y)'] - y_hat
df['Residuals'] = residuals
fig2 = px.histogram(df, x='Residuals', nbins=10, title='Histogram of Residuals',
                    labels={'Residuals': 'Residuals'})

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=('Crop Yield vs. Amount of Fertilizer',
                                    'Histogram of Residuals'))
for trace in fig1.data:
    fig.add_trace(trace, row=1, col=1)
for trace in fig2.data:
    fig.add_trace(trace, row=1, col=2)
fig.update_layout(title='Scatter Plot and Histogram of Residuals',
    xaxis_title='Amount of Fertilizer (kg)', yaxis_title='Crop Yield (tons)',
    xaxis2_title='Residuals', yaxis2_title='Frequency', showlegend=False)
fig.show()

Based on the screenshots you provided, we can analyze several assumptions of a simple linear regression model and identify which assumptions appear incompatible with the sample data:

### Assumptions of a Simple Linear Regression Model

1. **Linearity**:
   - **Assumption**: There is a linear relationship between the independent variable (x) and the dependent variable (y).
   - **Analysis**: From the scatter plot, we observe that as fertilizer usage increases, crop yield initially rises but then decreases, especially after fertilizer usage surpasses a certain point. This suggests that the data may not follow a linear relationship, violating the linearity assumption.

2. **Homoscedasticity**:
   - **Assumption**: The variance of the error terms is constant across the range of the independent variable.
   - **Analysis**: From the residual histogram, the residuals do not appear uniformly distributed, with certain areas showing a higher frequency of residuals than others. This indicates that the variance of the error terms may not be constant, violating the homoscedasticity assumption.

3. **Independence of Errors**:
   - **Assumption**: The error terms are independent of each other.
   - **Analysis**: From the provided image, we cannot directly observe the correlation between the error terms. However, if there is any pattern or periodicity in the data, this might suggest correlation between error terms.

4. **Normality of Errors**:
   - **Assumption**: The error terms are normally distributed.
   - **Analysis**: The residual histogram shows that the residuals may be skewed, with a heavier tail particularly for larger residuals. This may indicate that the error terms do not fully follow a normal distribution.

5. **No Perfect Multicollinearity**:
   - **Assumption**: In simple linear regression, as there is only one independent variable, this assumption is naturally satisfied because there are no other variables to be collinear with.

### Conclusion

Based on the images provided, the following assumptions seem incompatible with the sample data:

- **Linearity**: The nonlinear trend in the data suggests that the relationship between the independent and dependent variables may not be linear.
- **Homoscedasticity**: The uneven distribution of residuals suggests that the variance of the error terms may not be constant.
- **Normality of Errors**: The residual distribution may not follow a normal distribution, especially for larger residuals.

These analyses are based on visual observations of the images. More precise analysis would require statistical tests, such as tests for normality of residuals and homoscedasticity.


8. Specify a null hypothesis of "no linear association (on average)" in terms of the relevant parameter of the Simple Linear Regression model, and use the code below to characterize the evidence in the data relative to the null hypothesis and interpret your subsequent beliefs regarding the Old Faithful Geyser dataset.

In [ ]:
import plotly.express as px
import seaborn as sns
import statsmodels.api as sm
import pandas as pd

# The "Classic" Old Faithful Geyser dataset: ask a ChatBot for more details if desired
old_faithful = sns.load_dataset('geyser')

# Create a scatter plot with a Simple Linear Regression trendline
fig = px.scatter(old_faithful, x='waiting', y='duration',
                 title="Old Faithful Geyser Eruptions",
                 trendline='ols')#'lowess'

# Add a smoothed LOWESS Trendline to the scatter plot
lowess = sm.nonparametric.lowess  # Adjust 'frac' to change "smoothness bandwidth"
smoothed = lowess(old_faithful['duration'], old_faithful['waiting'], frac=0.25)
smoothed_df = pd.DataFrame(smoothed, columns=['waiting', 'smoothed_duration'])
fig.add_scatter(x=smoothed_df['waiting'], y=smoothed_df['smoothed_duration'],
                mode='lines', name='LOWESS Trendline')

fig.show() # USE `fig.show(renderer="png")` FOR ALL GitHub and MarkUs SUBMISSIONS

In [ ]:
import seaborn as sns
import statsmodels.formula.api as smf

# The "Classic" Old Faithful Geyser dataset
old_faithful = sns.load_dataset('geyser')
print(old_faithful)

linear_for_specification = 'duration ~ waiting'
model = smf.ols(linear_for_specification, data=old_faithful)
fitted_model = model.fit()
fitted_model.summary()

     duration  waiting   kind
0       3.600       79   long
1       1.800       54  short
2       3.333       74   long
3       2.283       62  short
4       4.533       85   long
..        ...      ...    ...
267     4.117       81   long
268     2.150       46  short
269     4.417       90   long
270     1.817       46  short
271     4.467       74   long

[272 rows x 3 columns]


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               duration   R-squared:                       0.811
Model:                            OLS   Adj. R-squared:                  0.811
Method:                 Least Squares   F-statistic:                     1162.
Date:                Wed, 06 Nov 2024   Prob (F-statistic):          8.13e-100
Time:                        12:08:10   Log-Likelihood:                -194.51
No. Observations:                 272   AIC:                             393.0
Df Residuals:                     270   BIC:                             400.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.8740      0.160    -11.702      0.000      -2.189      -1.559
waiting        0.0756      0.002     34.089      0.000       0.071       0.080
==============================================================================
Omnibus:                        4.133   Durbin-Watson:                   2.561
Prob(Omnibus):                  0.127   Jarque-Bera (JB):                3.173
Skew:                          -0.138   Prob(JB):                        0.205
Kurtosis:                       2.548   Cond. No.                         384.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

H
0
​
 :β
1
​
 =0 - There is no linear association between the 'waiting' time and the 'duration' of the eruptions (on average), where

β
1
​
  is the slope coefficient of the waiting variable in the regression model.


smf.ols prepares an OLS model where 'duration' is regressed on 'waiting'.
fit() fits the model to the data.
summary() provides a detailed summary of the regression results, including an estimate of

β
1
​
 , the standard error of this estimate, the t-statistic for testing

β
1
​
 =0, and the corresponding p-value.
Interpreting Results
Check the p-value associated with

β
1
​
 : p-value is 0, we have very strong evidence to reject the null hypothesis, concluding that there is a statistically significant linear relationship between 'waiting' time and 'duration'.
The coefficient for 'waiting' (

β
1
​
 ) indicates how much 'duration' changes for a one-minute increase in 'waiting'. A positive

β
1
​
  indicates a positive relationship; as waiting time increases, the duration of the eruption also increases.


9. As shown in the figure above, if the interval between the last geyser eruption and the current eruption exceeds approximately 63 minutes, the duration of the geyser eruption itself increases significantly. Therefore, in the figure below, we limit the dataset to only short waiting times. In the context of only short waiting times, is there evidence in the data that suggests a relationship between duration and waiting time similar to that in the full dataset? Using the code below, describe the evidence against the null hypothesis for cases where short waiting times are less than values of 62, 64, and 66.

In [ ]:
import plotly.express as px
import statsmodels.formula.api as smf


short_wait_limit = 62 # 64 # 66 #
short_wait = old_faithful.waiting < short_wait_limit

print(smf.ols('duration ~ waiting', data=old_faithful[short_wait]).fit().summary().tables[1])

# Create a scatter plot with a linear regression trendline
fig = px.scatter(old_faithful[short_wait], x='waiting', y='duration',
                 title="Old Faithful Geyser Eruptions for short wait times (<"+str(short_wait_limit)+")",
                 trendline='ols')

fig.show()

                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.6401      0.309      5.306      0.000       1.025       2.255
waiting        0.0069      0.006      1.188      0.238      -0.005       0.019


In [ ]:
import plotly.express as px
import statsmodels.formula.api as smf


short_wait_limit = 64 # 66 #
short_wait = old_faithful.waiting < short_wait_limit

print(smf.ols('duration ~ waiting', data=old_faithful[short_wait]).fit().summary().tables[1])

# Create a scatter plot with a linear regression trendline
fig = px.scatter(old_faithful[short_wait], x='waiting', y='duration',
                 title="Old Faithful Geyser Eruptions for short wait times (<"+str(short_wait_limit)+")",
                 trendline='ols')

fig.show()

                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.4140      0.288      4.915      0.000       0.842       1.986
waiting        0.0114      0.005      2.127      0.036       0.001       0.022


In [ ]:
import plotly.express as px
import statsmodels.formula.api as smf


short_wait_limit = 66 #
short_wait = old_faithful.waiting < short_wait_limit

print(smf.ols('duration ~ waiting', data=old_faithful[short_wait]).fit().summary().tables[1])

# Create a scatter plot with a linear regression trendline
fig = px.scatter(old_faithful[short_wait], x='waiting', y='duration',
                 title="Old Faithful Geyser Eruptions for short wait times (<"+str(short_wait_limit)+")",
                 trendline='ols')

fig.show()

                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.8619      0.327      2.637      0.010       0.213       1.511
waiting        0.0221      0.006      3.704      0.000       0.010       0.034


For the short wait limit of 62 minutes:

- **Waiting Coefficient (β₁):** 0.0069
- **P-value for Waiting:** p = 0.238

Since the p-value is above the common significance level (α = 0.05), we lack sufficient statistical evidence to reject the null hypothesis at this wait limit. This implies there is no statistically significant linear relationship between waiting time and eruption duration within this subset of data.

For the short wait limit of 64 minutes:

- **Waiting Coefficient (β₁):** 0.0114
- **P-value for Waiting:** p = 0.036

The p-value of 0.036 suggests moderate evidence to reject the null hypothesis, indicating a statistically significant linear relationship between waiting time and eruption duration for wait times under 64 minutes.

For the short wait limit of 66 minutes:

- **Waiting Coefficient (β₁):** 0.0221
- **P-value for Waiting:** p = 0.010

With a p-value of 0.01, there is strong evidence to reject the null hypothesis, showing a statistically significant linear relationship between waiting time and eruption duration for wait times under 66 minutes.

**Visual Analysis of Scatter Plots:**

- At 62 minutes, the data points show no clear upward trend, aligning with the statistical finding of a non-significant relationship.
- For 64 and 66 minutes, the trend lines exhibit a more evident positive slope, visually supporting the statistical results that show a significant relationship.

**Conclusions:**

- For short wait limits of 64 and 66 minutes, there is evidence of a significant relationship between eruption duration and waiting time, reflecting the trend in the overall dataset. This relationship becomes statistically significant as the wait limit increases.
- At a limit of 62 minutes, the relationship is not significant, suggesting that slightly longer wait times within this short-wait category may strengthen the relationship.

10. Let's now consider just the (n=160) long wait times (as specified in the code below), and write code to
create fitted Simple Linear Regression models for boostrap samples and collect and visualize the bootstrapped sampling distribution of the fitted slope coefficients of the fitted models;
simulate samples (of size n=160) from a Simple Linear Regression model that uses
,
,
 along with the values of waiting for
 to create simuations of
 and use these collect and visualize the sampling distribution of the fitted slope coefficients under a null hypothesis assumption of "no linear association (on average)"; then,
report if
 contained within a 95% bootstrapped confidence interval; and if the simulated p-value matches smf.ols('duration ~ waiting', data=old_faithful[long_wait]).fit().summary().tables[1]?

In [ ]:
import plotly.express as px

long_wait_limit = 71
long_wait = old_faithful.waiting > long_wait_limit

print(smf.ols('duration ~ waiting', data=old_faithful[long_wait]).fit().summary().tables[1])

# Create a scatter plot with a linear regression trendline
fig = px.scatter(old_faithful[long_wait], x='waiting', y='duration',
                 title="Old Faithful Geyser Eruptions for short wait times (<"+str(short_wait_limit)+")",
                 trendline='ols')

fig.show() # USE `fig.show(renderer="png")` FOR ALL GitHub and MarkUs SUBMISSIONS

                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.3520      0.476      7.049      0.000       2.413       4.291
waiting        0.0122      0.006      2.091      0.038       0.001       0.024


In [ ]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
from scipy import stats

# Assuming 'old_faithful' is a DataFrame containing the dataset
# For demonstration, I'm creating a sample dataset
np.random.seed(0)
old_faithful = pd.DataFrame({
    'waiting': np.random.normal(70, 12, 300),  # Simulated 'waiting' times around 70 with some variance
    'duration': np.random.normal(3.5, 1, 300)  # Simulated 'duration' of eruptions
})

# Filter for long wait times
long_wait_limit = 71
long_wait = old_faithful['waiting'] > long_wait_limit
data_long_wait = old_faithful[long_wait].sample(n=160, replace=True)

# Bootstrap samples and fit models
bootstrap_slopes = []
n_bootstraps = 1000
for _ in range(n_bootstraps):
    sample = data_long_wait.sample(n=160, replace=True)
    model = sm.OLS(sample['duration'], sm.add_constant(sample['waiting'])).fit()
    bootstrap_slopes.append(model.params[1])

# Simulate data under null hypothesis (no linear association)
simulated_slopes = []
for _ in range(n_bootstraps):
    simulated_duration = 1.65 + 0 * data_long_wait['waiting'] + np.random.normal(0, 0.37, size=160)
    model = sm.OLS(simulated_duration, sm.add_constant(data_long_wait['waiting'])).fit()
    simulated_slopes.append(model.params[1])


# Calculate 95% confidence interval for bootstrap slopes
conf_interval = np.percentile(bootstrap_slopes, [2.5, 97.5])

# Calculate the p-value for the simulated slopes
observed_slope = sm.OLS(data_long_wait['duration'], sm.add_constant(data_long_wait['waiting'])).fit().params[1]
p_value = np.mean(np.abs(simulated_slopes) >= np.abs(observed_slope))

# Output the results
print(f"95% Confidence Interval for Bootstrapped Slopes: {conf_interval}")
print(f"P-value for the Null Hypothesis: {p_value}")

# Plot the results
plt.figure(figsize=(12, 6))
plt.hist(bootstrap_slopes, bins=30, alpha=0.5, label='Bootstrapped Slopes')
plt.hist(simulated_slopes, bins=30, alpha=0.5, label='Simulated Slopes under Null', color='red')
plt.axvline(x=observed_slope, color='green', linestyle='--', label='Observed Slope')
plt.axvline(x=conf_interval[0], color='blue', linestyle='--', label='95% Confidence Interval')
plt.axvline(x=conf_interval[1], color='blue', linestyle='--')
plt.legend()
plt.show()


**Analysis of Results:**

**95% Confidence Interval:** The bootstrapped slopes' confidence interval, ranging from around -0.040 to 0.002, includes zero. This suggests that there is no statistically significant slope coefficient at the 5% level based on the bootstrapped samples.

**P-value:** The p-value for testing the null hypothesis ("no average linear association") is 0.0. This indicates that none of the slope values generated under the null hypothesis matched the extremity of the observed slope, providing strong evidence against the null hypothesis and suggesting a statistically significant relationship between 'waiting' and 'duration.'

**Observed Slope:** The observed slope is nearly zero (indicated by the green line almost intersecting the x-axis at zero), aligning with simulations under the null hypothesis and suggesting only a slight association.

**Interpretation:** While the 95% confidence interval includes zero, implying no significant relationship, the p-value indicates that the observed data strongly contradicts the null hypothesis of no relationship. This discrepancy may be due to the simulated slopes' narrow distribution compared to the variability seen in the bootstrapped slopes.

If the observed slope is close to zero and the p-value is still 0.0, this could imply that the null hypothesis simulation didn’t capture the observed data’s variability, or minor differences in model specifications impacted the results.

This analysis underscores the importance of examining both confidence intervals and p-values in context, as they may offer seemingly contradictory insights. Further investigation into both the real and simulated data distributions may help clarify these findings. Ensuring that assumptions like normality and error independence hold is also crucial for drawing robust conclusions from the regression analysis.




11. Since we've considered wait times of <64 "short" and wait times of >71 "long", let's instead just divide the data and insead call wait times of <68 "short" and otherwise just call them "long". Consider the Simple Linear Regression model specification using an indicator variable of the wait time length where we use
 (rather than
) (to refer to the "kind" or "katagory" or "kontrast") column (that you may have noticed was already a part) of the original dataset

and explain the "big picture" differences between this model specification and the previously considered model specifications and report the evidence against a null hypothesis of "no difference between groups "on average") for the new indicator variable based model
smf.ols('duration ~ waiting', data=old_faithful)
smf.ols('duration ~ waiting', data=old_faithful[short_wait])
smf.ols('duration ~ waiting', data=old_faithful[long_wait])

In [ ]:
from IPython.display import display

display(smf.ols('duration ~ C(kind, Treatment(reference="short"))', data=old_faithful).fit().summary().tables[1])

fig = px.box(old_faithful, x='kind', y='duration',
             title='duration ~ kind',
             category_orders={'kind': ['short', 'long']})
fig.show() # USE `fig.show(renderer="png")` FOR ALL GitHub and MarkUs SUBMISSIONS

<class 'statsmodels.iolib.table.SimpleTable'>

In [ ]:
import pandas as pd
import seaborn as sns
import statsmodels.formula.api as smf
import numpy as np

# Load the dataset
old_faithful = sns.load_dataset('geyser')

# Recategorize the wait times
old_faithful['kontrast'] = np.where(old_faithful['waiting'] < 68, 'short', 'long')

# Fit the overall model using an indicator variable
overall_model = smf.ols('duration ~ C(kontrast, Treatment(reference="short"))', data=old_faithful).fit()

# Fit the model for short wait times
short_wait = old_faithful['waiting'] < 68
model_short = smf.ols('duration ~ waiting', data=old_faithful[short_wait]).fit()

# Fit the model for long wait times
long_wait = old_faithful['waiting'] >= 68
model_long = smf.ols('duration ~ waiting', data=old_faithful[long_wait]).fit()

# Display the summary of the models
print("Overall Model (Indicator Variable):")
print(overall_model.summary())
print("\nModel for Short Wait Data:")
print(model_short.summary())
print("\nModel for Long Wait Data:")
print(model_long.summary())



Overall Model (Indicator Variable):
                            OLS Regression Results                            
Dep. Variable:               duration   R-squared:                       0.870
Model:                            OLS   Adj. R-squared:                  0.869
Method:                 Least Squares   F-statistic:                     1803.
Date:                Wed, 06 Nov 2024   Prob (F-statistic):          1.60e-121
Time:                        12:09:31   Log-Likelihood:                -144.19
No. Observations:                 272   AIC:                             292.4
Df Residuals:                     270   BIC:                             299.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                                                        coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------

The null hypothesis for this categorical model would state that there is no difference in eruption durations between "long" and "short" wait times ("on average"). Since p-value=0, we have very strong evidence against null hypothesis , indicating a statistically significant difference in duration between the two categories.



12.

In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go
from scipy import stats
import numpy as np

model_residuals = {
    '<br>Model 1:<br>All Data using slope': smf.ols('duration ~ waiting', data=old_faithful).fit().resid,
    '<br>Model 2:<br>Short Wait Data': smf.ols('duration ~ waiting', data=old_faithful[short_wait]).fit().resid,
    '<br>Model 3:<br>Long Wait Data': smf.ols('duration ~ waiting', data=old_faithful[long_wait]).fit().resid,
    '<br>Model 4:<br>All Data using indicator': smf.ols('duration ~ C(kind, Treatment(reference="short"))', data=old_faithful).fit().resid
}

fig = make_subplots(rows=2, cols=2, subplot_titles=list(model_residuals.keys()))
for i, (title, resid) in enumerate(model_residuals.items()):

    if i == 1:  # Apply different bins only to the second histogram (index 1)
        bin_size = dict(start=-1.9, end=1.9, size=0.2)
    else:
        bin_size = dict(start=-1.95, end=1.95, size=0.3)

    fig.add_trace(go.Histogram(x=resid, name=title, xbins=bin_size, histnorm='probability density'),
                  row=int(i/2)+1, col=(i%2)+1)
    fig.update_xaxes(title_text="n="+str(len(resid)), row=int(i/2)+1, col=(i%2)+1)

    normal_range = np.arange(-3*resid.std(),3*resid.std(),0.01)
    fig.add_trace(go.Scatter(x=normal_range, mode='lines', opacity=0.5,
                             y=stats.norm(loc=0, scale=resid.std()).pdf(normal_range),
                             line=dict(color='black', dash='dot', width=2),
                             name='Normal Distribution<br>(99.7% of its area)'),
                  row=int(i/2)+1, col=(i%2)+1)

fig.update_layout(title_text='Histograms of Residuals from Different Models')
fig.update_xaxes(range=[-2,2])
fig.show() # USE `fig.show(renderer="png")` FOR ALL GitHub and MarkUs SUBMISSIONS

The histograms for the four linear regression models indicate different levels of plausibility regarding the normality of their error terms:

1. **Model 1: All Data Using Slope** - The histogram suggests a near-normal distribution, though there are minor deviations at the tails. While the normal curve aligns with some sections of the histogram, it diverges at the tails, which could indicate issues like outliers or varying error variances.

2. **Model 2: Short Wait Data** - This histogram shows a clear departure from normality, skewed towards positive residuals. The right tail is significantly heavier, and the normal curve does not fit well across the distribution, implying that residuals deviate from normality.

3. **Model 3: Long Wait Data** - The histogram here resembles a normal distribution more closely than the others, following the normal curve fairly well. This suggests that the normality assumption for error terms is reasonable for this subset.

4. **Model 4: All Data Using Indicator** - This histogram poorly matches a normal distribution, showing a pronounced right tail and a peak shifted to negative residuals. The normal curve does not align well, indicating non-normality in the residuals.

Overall, **Model 3 (Long Wait Data)** appears to best satisfy the normality assumption for error terms, aligning with classical linear regression requirements. The other models exhibit skewness and kurtosis, suggesting that transformations, robust regression, or non-parametric methods may be suitable depending on the data and analysis goals.

13.Task Explanation
Permutation Testing: This involves randomly shuffling the 'kind' labels among the data and then recalculating the difference in means between the groups defined by these shuffled labels. This simulates a sampling distribution under the null hypothesis that the group labels (and thus the group membership) have no effect on the measured durations.

Bootstrap Confidence Interval: Here, you'll repeatedly sample with replacement from each group separately, calculate the mean for each resample, and then compute the difference in these means. This will help you construct a 95% confidence interval for the difference in means between the groups.



In [ ]:
from scipy import stats
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt

# Assuming 'old_faithful' is a pandas DataFrame containing the Old Faithful data
# with a column 'waiting' for waiting times and 'duration' for eruption durations.

# Define 'short' and 'long' wait groups based on a new threshold
old_faithful['kind'] = np.where(old_faithful['waiting'] <= 68, 'short', 'long')

# Perform permutation test
def permutation_test(data, n_permutations=10000):
    observed_diff = data.groupby('kind')['duration'].mean().diff().iloc[-1]
    permutation_diffs = []

    for _ in range(n_permutations):
        # Shuffle the 'kind' labels
        shuffled_kind = data['kind'].sample(frac=1, replace=False).reset_index(drop=True)
        permuted_data = data.assign(kind=shuffled_kind)
        mean_diff = permuted_data.groupby('kind')['duration'].mean().diff().iloc[-1]
        permutation_diffs.append(mean_diff)

    p_value = np.mean(np.abs(permutation_diffs) >= np.abs(observed_diff))
    return observed_diff, permutation_diffs, p_value

obs_diff, perm_diffs, p_val = permutation_test(old_faithful)

# Bootstrap confidence intervals
def bootstrap_CI(data, n_bootstraps=10000):
    boot_diffs = []

    for _ in range(n_bootstraps):
        boot_sample = data.sample(n=len(data), replace=True)
        mean_diff = boot_sample.groupby('kind')['duration'].mean().diff().iloc[-1]
        boot_diffs.append(mean_diff)

    ci_lower, ci_upper = np.percentile(boot_diffs, [2.5, 97.5])
    return boot_diffs, ci_lower, ci_upper

boot_diffs, ci_lower, ci_upper = bootstrap_CI(old_faithful)


In [ ]:
print(f"Observed difference: {obs_diff}, P-value: {p_val}")
print(f"95% CI for bootstrapped differences: [{ci_lower}, {ci_upper}]")

Observed difference: -2.1969790400092646, P-value: 0.0
95% CI for bootstrapped differences: [-2.2961314038298757, -2.0910806048310064]


In response to Question 13, two distinct statistical techniques are used to test the null hypothesis that there is no difference in wait times between short and long durations in the Old Faithful Geyser dataset. Though both methods address the same hypothesis, they differ in their resampling and testing approaches:

### A. Permutation Test
The permutation test involves "shuffling" the labels (short and long) and recalculating the mean difference between the shuffled groups. This process is repeated multiple times (e.g., 1,000 or more), creating a distribution of mean differences based on the assumption of no difference between the groups. This test directly assesses the hypothesis by comparing the observed mean difference to the distribution generated through random shuffling, determining if the observed difference is statistically significant.

### B. Bootstrap Confidence Interval
The bootstrap method involves repeatedly sampling (with replacement) within each group to calculate the mean for each resample. The differences in these bootstrapped means form a sampling distribution. A 95% confidence interval is then created from this distribution to check if it includes zero, which would imply no significant difference between the groups under the null hypothesis.

### Comparisons and Contrasts:

- **Methodology**: The permutation test simulates the null hypothesis by breaking any association between groups and outcome through reshuffling labels, while bootstrapping resamples within each group to estimate sampling variability and construct a confidence interval.

- **Assumptions**: Permutation tests are non-parametric, relying on empirical data alone without assuming a specific data distribution, making them suitable for non-normal data. Bootstrap methods assume that the sample is representative of the population and that resampling reflects the uncertainty in the estimate.

- **Application to Indicator Variables**: Both methods can also be used with models that include indicator variables (as in Question 11), where these variables distinguish groups for comparison. Indicator variable models handle group differences directly by estimating separate parameters, which can then be tested for significance within the model.

- **Similarities and Differences**: Both approaches aim to test the same hypothesis but from different angles. The permutation test directly evaluates the probability of observing the data under the null hypothesis, while bootstrapping assesses the variability of the mean difference and checks if zero lies within this interval. Each method provides valuable perspectives on the data.

These approaches complement indicator variable regression models by offering distribution-based hypothesis testing, which can be especially helpful when assumptions of normal regression (such as normality of residuals) may not be satisfied.

Your results demonstrate a strong statistical significance in the difference between short and long wait times. Here’s an explanation of what your findings suggest:

1. **Observed Difference**:
   - The observed difference of about -2.197 indicates a notable reduction in the average wait time when comparing long to short intervals. This figure represents the actual mean difference found in your data.

2. **P-value**:
   - With a P-value of 0.0, the observed difference is highly significant, assuming the null hypothesis (which states there's no difference in duration between the short and long groups). Such a low P-value supports rejecting the null hypothesis, indicating that the difference isn’t due to random chance.

3. **95% Confidence Interval**:
   - The confidence interval of [-2.296, -2.091] excludes zero and is entirely negative, reinforcing that long wait times are associated with significantly shorter durations. This consistency across bootstrapped samples adds further credibility to the finding.

These results provide robust evidence of a statistically significant difference in durations between the two groups. The confidence interval’s exclusion of zero strengthens the conclusion, giving substantial support against the null hypothesis. If relevant to your research question, these findings could provide strong support for your study or report on the effect of wait times on duration in the Old Faithful dataset.

14.Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Yes, and I work well with chatbots.